py/objlist.c

```diff
@@ -156,7 +156,7 @@ STATIC mp_obj_t list_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
         // delete
 #if MICROPY_PY_BUILTINS_SLICE
         if (MP_OBJ_IS_TYPE(index, &mp_type_slice)) {
-            mp_obj_list_t *self = self_in;
+            mp_obj_list_t *self = MP_OBJ_CAST(self_in);
```

Comment:
To start with such a meta-python doesn't have to be anything more than the current uPy source code which is preprocessed, with the preprocessing doing almost nothing to begin with.  Then we add things as needed.  Eg we can resolve #704 by translating all external API names in the processor.  And we can add a first argument to all functions that is the global state.  Sooner or later we need to do these things and a preprocessor is the only way I can think of doing it properly.